The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9904706946/job/27362696379) [(plain)](https://github.com/rust-lang/rust/commit/be13cc55565a425757b69eb0c40a90206e431ed3/checks/27362696379/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

---- [ui] tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs stdout ----
diff of stderr:

13 LL |     arg.map(|v| &**v)
15 
- error: aborting due to 1 previous error
+ error[E0308]: mismatched types
17   --> $DIR/transforming-option-ref-issue-127545.rs:9:19
17   --> $DIR/transforming-option-ref-issue-127545.rs:9:19
18    |
19 LL |     arg.unwrap_or(&[])

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mismatched_types/transforming-option-ref-issue-127545/transforming-option-ref-issue-127545.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args mismatched_types/transforming-option-ref-issue-127545.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mismatched_types/transforming-option-ref-issue-127545" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/mismatched_types/transforming-option-ref-issue-127545/auxiliary"
--- stderr -------------------------------
error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs:5:5
   |
   |
LL | pub fn foo(arg: Option<&Vec<i32>>) -> Option<&[i32]> {
   |                                       -------------- expected `Option<&[i32]>` because of return type
LL |     arg //~ ERROR 5:5: 5:8: mismatched types [E0308]
   |     ^^^ expected `Option<&[i32]>`, found `Option<&Vec<i32>>`
   = note: expected enum `Option<&[i32]>`
              found enum `Option<&Vec<i32>>`
              found enum `Option<&Vec<i32>>`
help: try using `.map(|v| &**v)` to convert `Option<&Vec<i32>>` to `Option<&[i32]>`
   |
LL |     arg.map(|v| &**v) //~ ERROR 5:5: 5:8: mismatched types [E0308]

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs:9:19
   |
   |
LL |     arg.unwrap_or(&[]) //~ ERROR 9:19: 9:22: mismatched types [E0308]
   |         --------- ^^^ expected `&Vec<i32>`, found `&[_; 0]`
   |         arguments to this method are incorrect
   |
   = note: expected reference `&Vec<i32>`
   = note: expected reference `&Vec<i32>`
              found reference `&[_; 0]`
help: the return type of this call is `&[_; 0]` due to the type of the argument passed
   |
   |
LL |     arg.unwrap_or(&[]) //~ ERROR 9:19: 9:22: mismatched types [E0308]
   |     ^^^^^^^^^^^^^^---^
   |                   this argument influences the return type of `unwrap_or`
note: method defined here
  --> /rustc/FAKE_PREFIX/library/core/src/option.rs:983:12
help: use `Option::map_or` to deref inner value of `Option`
help: use `Option::map_or` to deref inner value of `Option`
   |
LL |     arg.map_or(&[], |v| v) //~ ERROR 9:19: 9:22: mismatched types [E0308]

error[E0308]: mismatched types
##[error]  --> /checkout/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs:13:19
   |
   |
LL |     arg.unwrap_or(v) //~ ERROR 13:19: 13:20: mismatched types [E0308]
   |         --------- ^ expected `&Vec<i32>`, found `&[i32]`
   |         arguments to this method are incorrect
   |
   = note: expected reference `&Vec<i32>`
              found reference `&'a [i32]`
              found reference `&'a [i32]`
help: the return type of this call is `&'a [i32]` due to the type of the argument passed
   |
   |
LL |     arg.unwrap_or(v) //~ ERROR 13:19: 13:20: mismatched types [E0308]
   |     ^^^^^^^^^^^^^^-^
   |                   this argument influences the return type of `unwrap_or`
note: method defined here
  --> /rustc/FAKE_PREFIX/library/core/src/option.rs:983:12
help: use `Option::map_or` to deref inner value of `Option`
help: use `Option::map_or` to deref inner value of `Option`
   |
LL |     arg.map_or(v, |v| v) //~ ERROR 13:19: 13:20: mismatched types [E0308]

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0308`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -1451,9 +1451,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let ty::Adt(callee_adt, _) = callee_ty.peel_refs().kind() else {
             return;
         };
-        if !self.tcx.is_diagnostic_item(sym::Option, callee_adt.did()) {
+        let adt_name = if self.tcx.is_diagnostic_item(sym::Option, callee_adt.did()) {
+            "Option"
+        } else if self.tcx.is_diagnostic_item(sym::Result, callee_adt.did()) {
+            "Result"
+        } else {
             return;
-        }
+        };
 
         if call_ident.map_or(true, |ident| ident.name != sym::unwrap_or) {
             return;
@@ -1484,14 +1488,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Ok(snip) => (snip, Applicability::MachineApplicable),
                 Err(_) => ("/* _ */".to_owned(), Applicability::MaybeIncorrect),
             };
-        let sugg = &format!("map_or({provided_snip}, |v| v)");
-        err.span_suggestion_verbose(
-            error_span,
-            "use `Option::map_or` to deref inner value of `Option`",
-            sugg,
-            applicability,
-        );
-        return;
+        let sugg = format!("map_or({provided_snip}, |v| v)");
+        let msg = format!("use `{adt_name}::map_or` to deref inner value of `{adt_name}`");
+        err.span_suggestion_verbose(error_span, msg, sugg, applicability);
     }
 
     /// Suggest wrapping the block in square brackets instead of curly braces

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs
@@ -1459,7 +1459,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             return;
         };
 
-        if call_ident.map_or(true, |ident| ident.name != sym::unwrap_or) {
+        let Some(call_ident) = call_ident else {
+            return;
+        };
+        if call_ident.name != sym::unwrap_or {
             return;
         }
 
@@ -1483,14 +1486,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         if !self.can_coerce(expected_ty, dummy_ty) {
             return;
         }
-        let (provided_snip, applicability) =
-            match self.tcx.sess.source_map().span_to_snippet(provided_expr.span) {
-                Ok(snip) => (snip, Applicability::MachineApplicable),
-                Err(_) => ("/* _ */".to_owned(), Applicability::MaybeIncorrect),
-            };
-        let sugg = format!("map_or({provided_snip}, |v| v)");
         let msg = format!("use `{adt_name}::map_or` to deref inner value of `{adt_name}`");
-        err.span_suggestion_verbose(error_span, msg, sugg, applicability);
+        err.multipart_suggestion_verbose(
+            msg,
+            vec![
+                (call_ident.span, "map_or".to_owned()),
+                (provided_expr.span.shrink_to_hi(), ", |v| v".to_owned()),
+            ],
+            Applicability::MachineApplicable,
+        );
     }
 
     /// Suggest wrapping the block in square brackets instead of curly braces

--- a/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs
+++ b/tests/ui/mismatched_types/transforming-option-ref-issue-127545.rs
@@ -12,3 +12,7 @@ pub fn bar(arg: Option<&Vec<i32>>) -> &[i32] {
 pub fn barzz<'a>(arg: Option<&'a Vec<i32>>, v: &'a [i32]) -> &'a [i32] {
     arg.unwrap_or(v) //~ ERROR 13:19: 13:20: mismatched types [E0308]
 }
+
+pub fn convert_result(arg: Result<&Vec<i32>, ()>) -> &[i32] {
+    arg.unwrap_or(&[]) //~ ERROR 17:19: 17:22: mismatched types [E0308]
+}

--- a/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
+++ b/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
@@ -35,7 +35,7 @@ note: method defined here
 help: use `Option::map_or` to deref inner value of `Option`
    |
 LL |     arg.map_or(&[], |v| v)
-   |         ~~~~~~~~~~~~~~~~~~
+   |         ~~~~~~    +++++++
 
 error[E0308]: mismatched types
   --> $DIR/transforming-option-ref-issue-127545.rs:13:19
@@ -59,7 +59,7 @@ note: method defined here
 help: use `Option::map_or` to deref inner value of `Option`
    |
 LL |     arg.map_or(v, |v| v)
-   |         ~~~~~~~~~~~~~~~~
+   |         ~~~~~~  +++++++
 
 error[E0308]: mismatched types
   --> $DIR/transforming-option-ref-issue-127545.rs:17:19
@@ -83,7 +83,7 @@ note: method defined here
 help: use `Result::map_or` to deref inner value of `Result`
    |
 LL |     arg.map_or(&[], |v| v)
-   |         ~~~~~~~~~~~~~~~~~~
+   |         ~~~~~~    +++++++
 
 error: aborting due to 4 previous errors
 

--- a/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
+++ b/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
@@ -80,6 +80,10 @@ LL |     arg.unwrap_or(&[])
    |                   this argument influences the return type of `unwrap_or`
 note: method defined here
   --> $SRC_DIR/core/src/result.rs:LL:COL
+help: use `Result::map_or` to deref inner value of `Result`
+   |
+LL |     arg.map_or(&[], |v| v)
+   |         ~~~~~~~~~~~~~~~~~~
 
 error: aborting due to 4 previous errors
 

--- a/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
+++ b/tests/ui/mismatched_types/transforming-option-ref-issue-127545.stderr
@@ -61,6 +61,26 @@ help: use `Option::map_or` to deref inner value of `Option`
 LL |     arg.map_or(v, |v| v)
    |         ~~~~~~~~~~~~~~~~
 
-error: aborting due to 3 previous errors
+error[E0308]: mismatched types
+  --> $DIR/transforming-option-ref-issue-127545.rs:17:19
+   |
+LL |     arg.unwrap_or(&[])
+   |         --------- ^^^ expected `&Vec<i32>`, found `&[_; 0]`
+   |         |
+   |         arguments to this method are incorrect
+   |
+   = note: expected reference `&Vec<i32>`
+              found reference `&[_; 0]`
+help: the return type of this call is `&[_; 0]` due to the type of the argument passed
+  --> $DIR/transforming-option-ref-issue-127545.rs:17:5
+   |
+LL |     arg.unwrap_or(&[])
+   |     ^^^^^^^^^^^^^^---^
+   |                   |
+   |                   this argument influences the return type of `unwrap_or`
+note: method defined here
+  --> $SRC_DIR/core/src/result.rs:LL:COL
+
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

